ts/metrics/system_metrics.py

```diff
@@ -49,74 +51,98 @@ def disk_available():
     system_metrics.append(Metric("DiskAvailable", data, "GB", dimension))
 
 
-def gpu_utilization(num_of_gpu):
+def collect_gpu_metrics(num_of_gpus):
     """
-    Collect gpu metrics.
-
-    :param num_of_gpu:
+    Collect GPU metrics. Supports NVIDIA and AMD GPUs.
+    :param num_of_gpus: Total number of available GPUs.
     :return:
     """
-    if num_of_gpu <= 0:
-        return
-
-    # pylint: disable=wrong-import-position
-    # pylint: disable=import-outside-toplevel
-    import nvgpu
-    import pynvml
-    from nvgpu import list_gpus
+    for gpu_index in range(num_of_gpus):
+        if torch.version.cuda:
+            free, total = torch.cuda.mem_get_info(gpu_index)
+            mem_used = (total - free) // 1024**2
+            gpu_mem_utilization = torch.cuda.memory_usage(gpu_index)
+            gpu_utilization = torch.cuda.utilization(gpu_index)
+        elif torch.version.hip:
+            # There is currently a bug in
+            # https://github.com/pytorch/pytorch/blob/838958de94ed3b9021ddb395fe3e7ed22a60b06c/torch/cuda/__init__.py#L1171
+            # which does not capture the rate/percentage correctly.
+            # Otherwise same methods could be used.
+            # https://rocm.docs.amd.com/projects/amdsmi/en/latest/how-to/using-amdsmi-for-python.html#amdsmi-get-gpu-activity
+            import amdsmi
+
+            try:
+                amdsmi.amdsmi_init()
+
+                handle = amdsmi.amdsmi_get_processor_handles()[gpu_index]
+                mem_used = amdsmi.amdsmi_get_gpu_vram_usage(handle)["vram_used"]
```

Comment:
I believe that torch.cuda.mem_get_info should work fine on our systems if we want to follow the same approach.

Reply:
@jataylo thank you for the comment! I am not sure if we can do that yet, see below an example when using torch2.4.1+rocm6.1:
```
(venv) root@6f6ab1e7f4fb:/workspaces/torch-serve-amd# rocm-smi


========================================= ROCm System Management Interface =========================================
=================================================== Concise Info ===================================================
Device  Node  IDs              Temp    Power  Partitions          SCLK    MCLK     Fan  Perf  PwrCap  VRAM%  GPU%  
              (DID,     GUID)  (Edge)  (Avg)  (Mem, Compute, ID)                                                   
====================================================================================================================
0       4     0x740c,   11743  40.0°C  92.0W  N/A, N/A, 0         800Mhz  1600Mhz  0%   auto  560.0W  93%    0%    
1       5     0x740c,   61477  35.0°C  N/A    N/A, N/A, 0         800Mhz  1600Mhz  0%   auto  0.0W    93%    0%    
2       2     0x740c,   19109  36.0°C  94.0W  N/A, N/A, 0         800Mhz  1600Mhz  0%   auto  560.0W  90%    0%    
3       3     0x740c,   27916  38.0°C  N/A    N/A, N/A, 0         800Mhz  1600Mhz  0%   auto  0.0W    93%    0%    
4       8     0x740c,   28341  39.0°C  93.0W  N/A, N/A, 0         800Mhz  1600Mhz  0%   auto  560.0W  0%     0%    
5       9     0x740c,   30122  43.0°C  N/A    N/A, N/A, 0         800Mhz  1600Mhz  0%   auto  0.0W    0%     0%    
6       6     0x740c,   9668   33.0°C  94.0W  N/A, N/A, 0         800Mhz  1600Mhz  0%   auto  560.0W  0%     0%    
7       7     0x740c,   3258   31.0°C  N/A    N/A, N/A, 0         800Mhz  1600Mhz  0%   auto  0.0W    0%     0%    
====================================================================================================================
=============================================== End of ROCm SMI Log ================================================
(venv) root@6f6ab1e7f4fb:/workspaces/torch-serve-amd# python -c "import amdsmi; import torch; print(*[(i, amdsmi.amdsmi_get_gpu_vram_usage(amdsmi.amdsmi_get_processor_handles()[i])) for i in range(torch.cuda.device_count())], sep='\n');"
(0, {'vram_total': 65501, 'vram_used': 61497})
(1, {'vram_total': 65501, 'vram_used': 61497})
(2, {'vram_total': 65501, 'vram_used': 59305})
(3, {'vram_total': 65501, 'vram_used': 61497})
(4, {'vram_total': 65501, 'vram_used': 13})
(5, {'vram_total': 65501, 'vram_used': 13})
(6, {'vram_total': 65501, 'vram_used': 13})
(7, {'vram_total': 65501, 'vram_used': 13})
(venv) root@6f6ab1e7f4fb:/workspaces/torch-serve-amd# python -c "import torch; print(*[(i, torch.cuda.mem_get_info(i)) for i in range(torch.cuda.device_count())], sep='\n');"
(0, (6444548096, 68702699520))
(1, (4242538496, 68702699520))
(2, (4242538496, 68702699520))
(3, (4242538496, 68702699520))
(4, (68591550464, 68702699520))
(5, (68591550464, 68702699520))
(6, (68591550464, 68702699520))
(7, (68591550464, 68702699520))
```

Note the VRAM% usage for devices 0-3. Here the amdsmi and handle-based approach seems to provide correct numbers, but when using `torch.cuda.mem_get_info()` and accessing devices by index, the information does not seem to be correct (note that `mem_get_info()` returns `(free, total)` memory used).